File: 116-PopulatingNextRightPointersInEachNode/116-PopulatingNextRightPointersInEachNode.js
// Last updated: 4/29/2026, 10:34:34 AM
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
// 2022-11-18
// 23:23s
// O(n)
// O(n)
var connect = function(root) {
    if(!root) {
        return null;
    }
    let queue = [ root ];
    while(queue.length !== 0) {
        const nextQueue = [];
        for(const node of queue) {
            if(node.left) {
                nextQueue.push(node.left);
            }
            if(node.right) {
                nextQueue.push(node.right);
            }
        }
        for(let i = 0; i < nextQueue.length - 1; i++) {
            nextQueue[i].next = nextQueue[i+1];
        }
        queue = nextQueue;
    }
    return root;
};

File: 116-PopulatingNextRightPointersInEachNode/116-PopulatingNextRightPointersInEachNode.py
# Last updated: 4/29/2026, 10:34:30 AM
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        queue = [ root ]
        while len(queue) != 0:
            nextQueue = []
            for i in range(len(queue)):
                node = queue[i];
                if node.left:
                    nextQueue.append(node.left)
                if node.right:
                    nextQueue.append(node.right)
                if i > 0:
                    prev = queue[i-1]
                    prev.next = node
                if i == len(queue) - 1:
                    node.next = None
            queue = nextQueue
        return root
                    
            